File: src/lib/reportPdfExport.ts
import jsPDF from 'jspdf';
import autoTable from 'jspdf-autotable';
import { ComprehensiveReportResponse } from '@/api/services/ReportService';

const COLORS = {
  brand: [25, 119, 204] as [number, number, number],
  dark: [30, 41, 59] as [number, number, number],
  body: [71, 85, 105] as [number, number, number],
  muted: [148, 163, 184] as [number, number, number],
  lightBg: [248, 250, 252] as [number, number, number],
};

const fmt = (n: number) => n.toLocaleString('en-US', { minimumFractionDigits: 2, maximumFractionDigits: 2 });

const formatDate = (dateStr: string): string => {
  try {
    const d = new Date(dateStr);
    if (isNaN(d.getTime())) return dateStr;
    return d.toLocaleDateString('en-US', { year: 'numeric', month: 'short', day: 'numeric' });
  } catch {
    return dateStr;
  }
};

const STATUS_LABELS: Record<string, string> = {
  completed: 'Completed', scheduled: 'Scheduled', checked_in: 'Checked In',
  cancelled: 'Cancelled', no_show: 'No Show',
};

export interface ExportOptions {
  period: string;
  dateRange: { startDate: string; endDate: string };
  dispensaryName?: string;
  doctorName?: string;
}

// ── Shared helpers ──

const addPageFooters = (doc: jsPDF) => {
  const pageWidth = doc.internal.pageSize.width;
  const pageHeight = doc.internal.pageSize.height;
  const margin = 15;
  const pages = (doc as any).internal.getNumberOfPages();
  for (let i = 1; i <= pages; i++) {
    doc.setPage(i);
    doc.setFontSize(8);
    doc.setTextColor(...COLORS.muted);
    doc.text(`Generated on ${new Date().toLocaleString()}`, margin, pageHeight - 8);
    doc.text(`Page ${i} of ${pages}`, pageWidth - margin, pageHeight - 8, { align: 'right' });
  }
};

const drawHeader = (doc: jsPDF, title: string, options: ExportOptions): number => {
  const pageWidth = doc.internal.pageSize.width;
  const margin = 15;

  doc.setFillColor(...COLORS.brand);
  doc.rect(0, 0, pageWidth, 28, 'F');
  doc.setTextColor(255, 255, 255);
  doc.setFontSize(18);
  doc.setFont('helvetica', 'bold');
  doc.text(title, margin, 12);

  doc.setFontSize(10);
  doc.setFont('helvetica', 'normal');
  const periodLabel = options.period.charAt(0).toUpperCase() + options.period.slice(1);
  const start = formatDate(options.dateRange.startDate);
  const end = formatDate(options.dateRange.endDate);
  const dateLabel = start === end ? start : `${start} - ${end}`;
  doc.text(`${periodLabel} Report | ${dateLabel}`, margin, 22);

  if (options.dispensaryName) {
    doc.text(`Dispensary: ${options.dispensaryName}`, pageWidth - margin, 12, { align: 'right' });
  }
  if (options.doctorName) {
    doc.text(`Doctor: ${options.doctorName}`, pageWidth - margin, 22, { align: 'right' });
  }

  return 36; // y position after header
};

const drawSummarySection = (doc: jsPDF, data: ComprehensiveReportResponse, startY: number): number => {
  const margin = 15;
  const pageWidth = doc.internal.pageSize.width;
  let y = startY;

  // Booking summary
  doc.setTextColor(...COLORS.dark);
  doc.setFontSize(12);
  doc.setFont('helvetica', 'bold');
  doc.text('Booking Summary', margin, y);
  y += 8;

  const { summary, revenue } = data;
  const summaryRows = [
    [`Total: ${summary.total}`, `Completed: ${summary.completed}`, `Scheduled: ${summary.scheduled}`],
    [`Checked In: ${summary.checkedIn}`, `Cancelled: ${summary.cancelled}`, `No Show: ${summary.noShow}`],
  ];

  doc.setFontSize(10);
  doc.setFont('helvetica', 'normal');
  doc.setTextColor(...COLORS.body);
  summaryRows.forEach(row => {
    row.forEach((item, i) => {
      doc.text(item, margin + i * 90, y);
    });
    y += 7;
  });
  y += 5;

  // Revenue box
  doc.setFillColor(...COLORS.lightBg);
  doc.rect(margin, y, pageWidth - 2 * margin, 24, 'F');
  doc.setFontSize(11);
  doc.setFont('helvetica', 'bold');
  doc.setTextColor(...COLORS.brand);
  doc.text('Revenue Summary', margin + 5, y + 7);

  doc.setFontSize(9);
  doc.setFont('helvetica', 'normal');
  doc.setTextColor(...COLORS.body);
  const revItems = [
    `Total Fee: Rs. ${fmt(revenue.totalFee)}`,
    `Doctor Fee: Rs. ${fmt(revenue.doctorFee)}`,
    `Dispensary Fee: Rs. ${fmt(revenue.dispensaryFee)}`,
    `Commission: Rs. ${fmt(revenue.bookingCommission)}`,
    `Realized: Rs. ${fmt(revenue.realizedRevenue)}`,
  ];
  revItems.forEach((item, i) => {
    const col = i % 3;
    const row = Math.floor(i / 3);
    doc.text(item, margin + 5 + col * 90, y + 14 + row * 6);
  });
  y += 32;

  return y;
};

// ── Export: Bookings only ──

export const exportBookingsReportToPDF = (
  data: ComprehensiveReportResponse,
  options: ExportOptions
) => {
  const doc = new jsPDF('l', 'mm', 'a4');
  const pageHeight = doc.internal.pageSize.height;
  const margin = 15;

  let y = drawHeader(doc, 'MyClinic - Booking Details Report', options);
  y = drawSummarySection(doc, data, y);

  if (data.bookings.length > 0) {
    if (y + 40 > pageHeight - 20) { doc.addPage(); y = 20; }

    doc.setFontSize(12);
    doc.setFont('helvetica', 'bold');
    doc.setTextColor(...COLORS.dark);
    doc.text(`Booking Details (${data.bookings.length} records)`, margin, y);
    y += 6;

    autoTable(doc, {
      startY: y,
      margin: { left: margin, right: margin },
      head: [['#', 'Date', 'Time', 'Patient', 'Phone', 'Doctor', 'Dispensary', 'Status', 'Total Fee']],
      body: data.bookings.map((b, i) => [
        String(i + 1),
        formatDate(b.bookingDate),
        b.timeSlot || '',
        b.patientName || '',
        b.patientPhone || '',
        b.doctorName || '',
        b.dispensaryName || '',
        STATUS_LABELS[b.status] || b.status,
        b.fees?.totalFee?.toFixed(2) || '-',
      ]),
      styles: { fontSize: 8, cellPadding: 2 },
      headStyles: { fillColor: COLORS.brand, textColor: 255, fontStyle: 'bold' },
      alternateRowStyles: { fillColor: COLORS.lightBg },
      columnStyles: {
        0: { cellWidth: 10 },
        8: { halign: 'right' as const },
      },
    });
  }

  addPageFooters(doc);
  doc.save(`Booking_Details_${options.period}_${new Date().toISOString().split('T')[0]}.pdf`);
};

// ── Export: Revenue only ──

export const exportRevenueReportToPDF = (
  data: ComprehensiveReportResponse,
  options: ExportOptions
) => {
  const doc = new jsPDF('l', 'mm', 'a4');
  const pageHeight = doc.internal.pageSize.height;
  const margin = 15;

  let y = drawHeader(doc, 'MyClinic - Revenue Report', options);
  y = drawSummarySection(doc, data, y);

  if (data.topDoctors.length > 0) {
    if (y + 40 > pageHeight - 20) { doc.addPage(); y = 20; }

    doc.setFontSize(12);
    doc.setFont('helvetica', 'bold');
    doc.setTextColor(...COLORS.dark);
    doc.text('Revenue by Doctor', margin, y);
    y += 6;

    autoTable(doc, {
      startY: y,
      margin: { left: margin, right: margin },
      head: [['Doctor', 'Bookings', 'Total Fee', 'Doctor Fee', 'Dispensary Fee', 'Commission', 'CP Fee']],
      body: data.topDoctors.map(d => [
        d.doctorName,
        String(d.bookingCount),
        fmt(d.totalFee),
        fmt(d.doctorFee),
        fmt(d.dispensaryFee),
        fmt(d.bookingCommission),
        fmt(d.channelPartnerFee),
      ]),
      styles: { fontSize: 9, cellPadding: 2 },
      headStyles: { fillColor: COLORS.brand, textColor: 255, fontStyle: 'bold' },
      alternateRowStyles: { fillColor: COLORS.lightBg },
    });
  }

  addPageFooters(doc);
  doc.save(`Revenue_Report_${options.period}_${new Date().toISOString().split('T')[0]}.pdf`);
};

// ── Export: Full comprehensive (both) ──

export const exportComprehensiveReportToPDF = (
  data: ComprehensiveReportResponse,
  options: ExportOptions
) => {
  const doc = new jsPDF('l', 'mm', 'a4');
  const pageHeight = doc.internal.pageSize.height;
  const margin = 15;

  let y = drawHeader(doc, 'MyClinic - Comprehensive Report', options);
  y = drawSummarySection(doc, data, y);

  // Revenue by Doctor table
  if (data.topDoctors.length > 0) {
    if (y + 40 > pageHeight - 20) { doc.addPage(); y = 20; }

    doc.setFontSize(12);
    doc.setFont('helvetica', 'bold');
    doc.setTextColor(...COLORS.dark);
    doc.text('Revenue by Doctor', margin, y);
    y += 6;

    autoTable(doc, {
      startY: y,
      margin: { left: margin, right: margin },
      head: [['Doctor', 'Bookings', 'Total Fee', 'Doctor Fee', 'Dispensary Fee', 'Commission', 'CP Fee']],
      body: data.topDoctors.map(d => [
        d.doctorName,
        String(d.bookingCount),
        fmt(d.totalFee),
        fmt(d.doctorFee),
        fmt(d.dispensaryFee),
        fmt(d.bookingCommission),
        fmt(d.channelPartnerFee),
      ]),
      styles: { fontSize: 9, cellPadding: 2 },
      headStyles: { fillColor: COLORS.brand, textColor: 255, fontStyle: 'bold' },
      alternateRowStyles: { fillColor: COLORS.lightBg },
    });
    y = (doc as any).lastAutoTable.finalY + 10;
  }

  // Booking Details table
  if (data.bookings.length > 0) {
    if (y + 40 > pageHeight - 20) { doc.addPage(); y = 20; }

    doc.setFontSize(12);
    doc.setFont('helvetica', 'bold');
    doc.setTextColor(...COLORS.dark);
    doc.text(`Booking Details (${data.bookings.length} records)`, margin, y);
    y += 6;

    autoTable(doc, {
      startY: y,
      margin: { left: margin, right: margin },
      head: [['#', 'Date', 'Time', 'Patient', 'Phone', 'Doctor', 'Dispensary', 'Status', 'Total Fee']],
      body: data.bookings.map((b, i) => [
        String(i + 1),
        formatDate(b.bookingDate),
        b.timeSlot || '',
        b.patientName || '',
        b.patientPhone || '',
        b.doctorName || '',
        b.dispensaryName || '',
        STATUS_LABELS[b.status] || b.status,
        b.fees?.totalFee?.toFixed(2) || '-',
      ]),
      styles: { fontSize: 8, cellPadding: 2 },
      headStyles: { fillColor: COLORS.brand, textColor: 255, fontStyle: 'bold' },
      alternateRowStyles: { fillColor: COLORS.lightBg },
      columnStyles: {
        0: { cellWidth: 10 },
        8: { halign: 'right' as const },
      },
    });
  }

  addPageFooters(doc);
  doc.save(`Comprehensive_Report_${options.period}_${new Date().toISOString().split('T')[0]}.pdf`);
};

File: myclinic-dispensary-app/lib/utils/report_pdf_export.dart
import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;
import 'package:printing/printing.dart';
import 'package:intl/intl.dart';

/// Colors matching the web app's PDF export
const _brand = PdfColor.fromInt(0xFF1977CC);
const _dark = PdfColor.fromInt(0xFF1E293B);
const _body = PdfColor.fromInt(0xFF475569);
const _muted = PdfColor.fromInt(0xFF94A3B8);
const _lightBg = PdfColor.fromInt(0xFFF8FAFC);
const _white = PdfColors.white;

String _fmt(dynamic val) {
  if (val == null) return '0.00';
  final n = (val is num) ? val.toDouble() : double.tryParse('$val') ?? 0;
  return n.toStringAsFixed(2);
}

String _fmtDate(dynamic val) {
  if (val == null) return 'N/A';
  try {
    return DateFormat('MMM dd, yyyy').format(DateTime.parse('$val').toLocal());
  } catch (_) {
    return '$val';
  }
}

String _statusLabel(String status) {
  switch (status) {
    case 'completed':
      return 'Completed';
    case 'scheduled':
      return 'Scheduled';
    case 'checked_in':
      return 'Checked In';
    case 'cancelled':
      return 'Cancelled';
    case 'no_show':
      return 'No Show';
    default:
      return status;
  }
}

/// Export comprehensive report (both bookings + revenue) as PDF
Future<void> exportComprehensiveReportToPDF({
  required Map<String, dynamic> data,
  required String period,
  required String startDate,
  required String endDate,
  String? dispensaryName,
  String? doctorName,
}) async {
  final pdf = pw.Document();
  final summary = data['summary'] as Map<String, dynamic>? ?? {};
  final revenue = data['revenue'] as Map<String, dynamic>? ?? {};
  final bookings = data['bookings'] as List? ?? [];
  final topDoctors = data['topDoctors'] as List? ?? [];
  final dateRange = '${_fmtDate(startDate)} - ${_fmtDate(endDate)}';

  pdf.addPage(
    pw.MultiPage(
      pageFormat: PdfPageFormat.a4.landscape,
      margin: const pw.EdgeInsets.all(15 * PdfPageFormat.mm),
      header: (context) => _buildHeader(
        title: 'Comprehensive Report',
        period: period,
        dateRange: dateRange,
        dispensaryName: dispensaryName,
        doctorName: doctorName,
      ),
      footer: _buildFooter,
      build: (context) => [
        // Booking summary
        _buildSummarySection(summary),
        pw.SizedBox(height: 10),
        // Revenue summary
        _buildRevenueBox(revenue),
        pw.SizedBox(height: 14),
        // Revenue by doctor table
        if (topDoctors.isNotEmpty) ...[
          _buildSectionTitle('Revenue by Doctor (${topDoctors.length})'),
          pw.SizedBox(height: 4),
          _buildDoctorTable(topDoctors, revenue),
          pw.SizedBox(height: 14),
        ],
        // Booking details table
        if (bookings.isNotEmpty) ...[
          _buildSectionTitle('Booking Details (${bookings.length})'),
          pw.SizedBox(height: 4),
          _buildBookingsTable(bookings),
        ],
      ],
    ),
  );

  final fileName =
      'Comprehensive_Report_${period}_${DateFormat('yyyy-MM-dd').format(DateTime.now())}.pdf';
  await Printing.sharePdf(bytes: await pdf.save(), filename: fileName);
}

/// Export bookings only
Future<void> exportBookingsReportToPDF({
  required Map<String, dynamic> data,
  required String period,
  required String startDate,
  required String endDate,
  String? dispensaryName,
  String? doctorName,
}) async {
  final pdf = pw.Document();
  final summary = data['summary'] as Map<String, dynamic>? ?? {};
  final bookings = data['bookings'] as List? ?? [];
  final dateRange = '${_fmtDate(startDate)} - ${_fmtDate(endDate)}';

  pdf.addPage(
    pw.MultiPage(
      pageFormat: PdfPageFormat.a4.landscape,
      margin: const pw.EdgeInsets.all(15 * PdfPageFormat.mm),
      header: (context) => _buildHeader(
        title: 'Booking Details Report',
        period: period,
        dateRange: dateRange,
        dispensaryName: dispensaryName,
        doctorName: doctorName,
      ),
      footer: _buildFooter,
      build: (context) => [
        _buildSummarySection(summary),
        pw.SizedBox(height: 14),
        if (bookings.isNotEmpty) ...[
          _buildSectionTitle('Booking Details (${bookings.length})'),
          pw.SizedBox(height: 4),
          _buildBookingsTable(bookings),
        ],
      ],
    ),
  );

  final fileName =
      'Booking_Details_${period}_${DateFormat('yyyy-MM-dd').format(DateTime.now())}.pdf';
  await Printing.sharePdf(bytes: await pdf.save(), filename: fileName);
}

/// Export revenue only
Future<void> exportRevenueReportToPDF({
  required Map<String, dynamic> data,
  required String period,
  required String startDate,
  required String endDate,
  String? dispensaryName,
  String? doctorName,
}) async {
  final pdf = pw.Document();
  final revenue = data['revenue'] as Map<String, dynamic>? ?? {};
  final topDoctors = data['topDoctors'] as List? ?? [];
  final dateRange = '${_fmtDate(startDate)} - ${_fmtDate(endDate)}';

  pdf.addPage(
    pw.MultiPage(
      pageFormat: PdfPageFormat.a4.landscape,
      margin: const pw.EdgeInsets.all(15 * PdfPageFormat.mm),
      header: (context) => _buildHeader(
        title: 'Revenue Summary Report',
        period: period,
        dateRange: dateRange,
        dispensaryName: dispensaryName,
        doctorName: doctorName,
      ),
      footer: _buildFooter,
      build: (context) => [
        _buildRevenueBox(revenue),
        pw.SizedBox(height: 14),
        if (topDoctors.isNotEmpty) ...[
          _buildSectionTitle('Revenue by Doctor (${topDoctors.length})'),
          pw.SizedBox(height: 4),
          _buildDoctorTable(topDoctors, revenue),
        ],
      ],
    ),
  );

  final fileName =
      'Revenue_Report_${period}_${DateFormat('yyyy-MM-dd').format(DateTime.now())}.pdf';
  await Printing.sharePdf(bytes: await pdf.save(), filename: fileName);
}

// ─── Building Blocks ─────────────────────────────────────────

pw.Widget _buildHeader({
  required String title,
  required String period,
  required String dateRange,
  String? dispensaryName,
  String? doctorName,
}) {
  return pw.Container(
    width: double.infinity,
    padding: const pw.EdgeInsets.symmetric(horizontal: 14, vertical: 12),
    margin: const pw.EdgeInsets.only(bottom: 10),
    decoration: const pw.BoxDecoration(
      color: _brand,
      borderRadius: pw.BorderRadius.all(pw.Radius.circular(4)),
    ),
    child: pw.Row(
      mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
      crossAxisAlignment: pw.CrossAxisAlignment.start,
      children: [
        pw.Column(
          crossAxisAlignment: pw.CrossAxisAlignment.start,
          children: [
            pw.Text(title,
                style: pw.TextStyle(
                    color: _white,
                    fontSize: 16,
                    fontWeight: pw.FontWeight.bold)),
            pw.SizedBox(height: 3),
            pw.Text(
                '${period[0].toUpperCase()}${period.substring(1)} | $dateRange',
                style: const pw.TextStyle(color: _white, fontSize: 9)),
          ],
        ),
        pw.Column(
          crossAxisAlignment: pw.CrossAxisAlignment.end,
          children: [
            if (dispensaryName != null)
              pw.Text(dispensaryName,
                  style: const pw.TextStyle(color: _white, fontSize: 9)),
            if (doctorName != null)
              pw.Text(doctorName,
                  style: const pw.TextStyle(color: _white, fontSize: 9)),
          ],
        ),
      ],
    ),
  );
}

pw.Widget _buildFooter(pw.Context context) {
  return pw.Container(
    width: double.infinity,
    padding: const pw.EdgeInsets.only(top: 6),
    child: pw.Row(
      mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
      children: [
        pw.Text(
          'Generated: ${DateFormat('MMM dd, yyyy hh:mm a').format(DateTime.now())}',
          style: const pw.TextStyle(color: _muted, fontSize: 7),
        ),
        pw.Text(
          'Page ${context.pageNumber} of ${context.pagesCount}',
          style: const pw.TextStyle(color: _muted, fontSize: 7),
        ),
      ],
    ),
  );
}

pw.Widget _buildSummarySection(Map<String, dynamic> summary) {
  return pw.Row(
    mainAxisAlignment: pw.MainAxisAlignment.spaceAround,
    children: [
      _summaryItem('Total', '${summary['total'] ?? 0}'),
      _summaryItem('Completed', '${summary['completed'] ?? 0}'),
      _summaryItem('Scheduled', '${summary['scheduled'] ?? 0}'),
      _summaryItem('Checked In', '${summary['checkedIn'] ?? 0}'),
      _summaryItem('Cancelled', '${summary['cancelled'] ?? 0}'),
      _summaryItem('No Show', '${summary['noShow'] ?? 0}'),
    ],
  );
}

pw.Widget _summaryItem(String label, String value) {
  return pw.Column(
    children: [
      pw.Text(value,
          style: pw.TextStyle(
              fontSize: 14, fontWeight: pw.FontWeight.bold, color: _dark)),
      pw.SizedBox(height: 2),
      pw.Text(label, style: const pw.TextStyle(fontSize: 8, color: _body)),
    ],
  );
}

pw.Widget _buildRevenueBox(Map<String, dynamic> revenue) {
  return pw.Container(
    width: double.infinity,
    padding: const pw.EdgeInsets.all(10),
    decoration: pw.BoxDecoration(
      color: _lightBg,
      borderRadius: pw.BorderRadius.circular(4),
    ),
    child: pw.Column(
      crossAxisAlignment: pw.CrossAxisAlignment.start,
      children: [
        pw.Text('Revenue Summary',
            style: pw.TextStyle(
                fontSize: 10,
                fontWeight: pw.FontWeight.bold,
                color: _brand)),
        pw.SizedBox(height: 6),
        pw.Row(
          mainAxisAlignment: pw.MainAxisAlignment.spaceAround,
          children: [
            _revenueItem('Total Fee', 'Rs. ${_fmt(revenue['totalFee'])}'),
            _revenueItem('Doctor Fee', 'Rs. ${_fmt(revenue['doctorFee'])}'),
            _revenueItem(
                'Dispensary Fee', 'Rs. ${_fmt(revenue['dispensaryFee'])}'),
            _revenueItem(
                'Commission', 'Rs. ${_fmt(revenue['bookingCommission'])}'),
            _revenueItem('Realized', 'Rs. ${_fmt(revenue['realizedRevenue'])}'),
          ],
        ),
      ],
    ),
  );
}

pw.Widget _revenueItem(String label, String value) {
  return pw.Column(
    children: [
      pw.Text(value,
          style: pw.TextStyle(
              fontSize: 9, fontWeight: pw.FontWeight.bold, color: _dark)),
      pw.SizedBox(height: 1),
      pw.Text(label, style: const pw.TextStyle(fontSize: 7, color: _body)),
    ],
  );
}

pw.Widget _buildSectionTitle(String text) {
  return pw.Text(text,
      style: pw.TextStyle(
          fontSize: 11, fontWeight: pw.FontWeight.bold, color: _dark));
}

pw.Widget _buildBookingsTable(List bookings) {
  return pw.TableHelper.fromTextArray(
    headerStyle:
        pw.TextStyle(fontSize: 7, fontWeight: pw.FontWeight.bold, color: _white),
    headerDecoration: const pw.BoxDecoration(color: _brand),
    cellStyle: const pw.TextStyle(fontSize: 7, color: _body),
    cellPadding: const pw.EdgeInsets.symmetric(horizontal: 4, vertical: 3),
    headerPadding: const pw.EdgeInsets.symmetric(horizontal: 4, vertical: 4),
    oddRowDecoration: const pw.BoxDecoration(color: _lightBg),
    headerAlignments: {
      0: pw.Alignment.centerLeft,
      8: pw.Alignment.centerRight,
    },
    columnWidths: {
      0: const pw.FixedColumnWidth(18), // #
      1: const pw.FixedColumnWidth(55), // Date
      2: const pw.FixedColumnWidth(40), // Time
      3: const pw.FlexColumnWidth(2), // Patient
      4: const pw.FixedColumnWidth(55), // Phone
      5: const pw.FlexColumnWidth(2), // Doctor
      6: const pw.FlexColumnWidth(2), // Dispensary
      7: const pw.FixedColumnWidth(42), // Status
      8: const pw.FixedColumnWidth(45), // Fee
    },
    headers: ['#', 'Date', 'Time', 'Patient', 'Phone', 'Doctor', 'Dispensary', 'Status', 'Total Fee'],
    data: bookings.asMap().entries.map((entry) {
      final i = entry.key;
      final b = entry.value as Map<String, dynamic>;
      final fees = b['fees'] as Map<String, dynamic>? ?? {};
      return [
        '${i + 1}',
        _fmtDate(b['bookingDate']),
        b['timeSlot'] ?? '',
        b['patientName'] ?? 'N/A',
        b['patientPhone'] ?? '',
        b['doctorName'] ?? '',
        b['dispensaryName'] ?? '',
        _statusLabel(b['status'] ?? ''),
        'Rs.${_fmt(fees['totalFee'])}',
      ];
    }).toList(),
  );
}

pw.Widget _buildDoctorTable(
    List topDoctors, Map<String, dynamic> revenue) {
  final rows = topDoctors.map((d) {
    final doc = d as Map<String, dynamic>;
    return [
      doc['doctorName'] ?? 'N/A',
      '${doc['bookingCount'] ?? 0}',
      'Rs.${_fmt(doc['totalFee'])}',
      'Rs.${_fmt(doc['doctorFee'])}',
      'Rs.${_fmt(doc['dispensaryFee'])}',
      'Rs.${_fmt(doc['bookingCommission'])}',
      'Rs.${_fmt(doc['channelPartnerFee'])}',
    ];
  }).toList();

  // Add totals row
  rows.add([
    'TOTAL',
    '${revenue['totalBookings'] ?? topDoctors.fold(0, (sum, d) => sum + ((d as Map)['bookingCount'] ?? 0) as int)}',
    'Rs.${_fmt(revenue['totalFee'])}',
    'Rs.${_fmt(revenue['doctorFee'])}',
    'Rs.${_fmt(revenue['dispensaryFee'])}',
    'Rs.${_fmt(revenue['bookingCommission'])}',
    'Rs.${_fmt(revenue['channelPartnerFee'])}',
  ]);

  return pw.TableHelper.fromTextArray(
    headerStyle:
        pw.TextStyle(fontSize: 8, fontWeight: pw.FontWeight.bold, color: _white),
    headerDecoration: const pw.BoxDecoration(color: _brand),
    cellStyle: const pw.TextStyle(fontSize: 8, color: _body),
    cellPadding: const pw.EdgeInsets.symmetric(horizontal: 4, vertical: 3),
    headerPadding: const pw.EdgeInsets.symmetric(horizontal: 4, vertical: 4),
    oddRowDecoration: const pw.BoxDecoration(color: _lightBg),
    headers: [
      'Doctor',
      'Bookings',
      'Total Fee',
      'Doctor Fee',
      'Disp. Fee',
      'Commission',
      'CP Fee'
    ],
    data: rows,
  );
}
